# Profitable App Profiles for the App Store and Google Play 

Our aim in this project:

- find mobile app profiles that are profitable for the App Store and Google Play markets.
- goal is to analyze data to help developers understand what kinds of apps are likely to attract more users.

Used data sets:

- [A data set](https://www.kaggle.com/lava18/google-play-store-apps/home) containing data about approximately ten thousand Android apps from Google Play; the data was collected in August 2018.
- [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) containing data about approximately seven thousand iOS apps from the App Store; the data was collected in July 2017.

### Open Data Sets And Save As Lists
Using `openfile` function, open data sets and save as lists. Also, using `explore_data` function, print number of data sets, the number of cloumns, and a few example of the data

In [1]:
from csv import reader

def openfile(dataset, header=True):
    opened_file = open(dataset)
    read_file = reader(opened_file)
    apps_data = list(read_file)
    
    #if a header exists(true), return a header and dataset separately
    if header:
        return apps_data[0], apps_data[1:]
    else:
        return apps_da

ios_store = openfile('AppleStore.csv')
ios_header = ios_store[0]
ios_data = ios_store[1]

android_play = openfile('googleplaystore.csv')
android_header = android_play[0]
android_data = android_play[1]

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

# The App Store data set
print(ios_header)
print('\n')
explore_data(ios_data, 0, 3, True)
print('\n')

# The Google Play data set
print(android_header)
print('\n')
explore_data(android_data, 0, 3, True)


['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 7197
Number of columns: 17


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,

### Removing Inaccurate Data 
Remove an inaccurate android data row based on [Kaggle discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015)

In [3]:
#Based on Kaggle discussion, the data missing "Category" on row 10472
#row[10472] entry has missing 'Rating' and a column shift happened for next columns
print(android_data[10472])
print('\n')
print(android_header)
print('\n')
print('Number of rows:', len(android_data))
del android_play[1][10472]
#After eliminating row 10472
print('Number of rows:', len(android_data), '(Confirm that exactly one row of data was removed)')

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Number of rows: 10841
Number of rows: 10840 (Confirm that exactly one row of data was removed)


### Removing Duplicate Data
Check if data sets have duplicate entries. Print number of duplicates and a few examples to confirm.

Check to see if android data set contains duplicate data

In [4]:
duplicate_app = []
unique_app = []

for app in android_data:
    name = app[0]
    if name in unique_app:
        duplicate_app.append(name)
    else:
        unique_app.append(name)

print('Number of duplicate apps:', len(duplicate_app), '\n')
print('Example of duplicate apps:', duplicate_app[:3])

Number of duplicate apps: 1181 

Example of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business']


Build a criterion for removing the duplicates.
Rather than removing duplicates randomly, only keep the row with the highest number of reviews (the most recent data shoud be) and remove the other entries for any given app.

In [5]:
reviews_max = {}
for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews

print('Number of non-duplicate apps:', len(reviews_max))

Number of non-duplicate apps: 9659


Use the dictionary created above to remove the duplicate rows and create clean data set

In [6]:
android_clean = []
already_added = []

for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
        
print('Number of non-duplicate app in :', len(android_clean))
print("Confirm that number match with 'non-duplicate app' number") 

Number of non-duplicate app in : 9659
Confirm that number match with 'non-duplicate app' number


Check to see if ios data set contains duplicate data:

In [7]:
duplicate_app_ios = []
unique_app_ios = []

for app in ios_data:
    name = app[1]
    if name in unique_app_ios:
        duplicate_app_ios.append(name)
    else:
        unique_app_ios.append(name)

print('Number of duplicate apps:', len(duplicate_app_ios))
print('Number of unique apps:', len(unique_app_ios))
print('Example of duplicate apps:', duplicate_app_ios[:3])

Number of duplicate apps: 0
Number of unique apps: 7197
Example of duplicate apps: []


There is no duplicate apps in ios dataset.

### Removing Non-English Apps Data
First, create a function that takes in a string and returns False if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns True.

In [10]:
def is_eng_app(string):
    for char in string:
        if ord(char) > 127:
            return False
    return True

#function check
print(is_eng_app('Instagram'))
print(is_eng_app('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_eng_app('Docs To Go™ Free Office Suite'))
print(is_eng_app('Instachat 😜'))

True
False
False
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English.

In [11]:
def is_eng_app(string):
    non_eng_char = 0
    for char in string:
        if ord(char) > 127:
            non_eng_char += 1
              
    if non_eng_char > 3:
        return False
    else:
        return True
    
#function check
print(is_eng_app('Docs To Go™ Free Office Suite'))
print(is_eng_app('Instachat 😜'))
print(is_eng_app('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Use this new function to filter out non-English apps from both data sets.

In [12]:
android_english = []
ios_english = []

# Android
for app in android_clean:
    name = app[0]
    if is_eng_app(name):
        android_english.append(app)

# ios
for app in ios_clean:
    name = app[1]
    if is_eng_app(name):
        ios_english.append(app)
        
print('Check number of android data set:', len(android_english)) 
print('Check number of ios data set:', len(ios_english)) 

Check number of android data set: 9614
Check number of ios data set: 7197


### Isolating Free Apps
So far in the data cleaning process, we:

- Removed inaccurate data
- Removed duplicate app entries
- Removed non-English apps

As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis.

In [13]:
def isolate_free_app(dataset, index, ans):
    free_apps = []
    for app in dataset:
        price = app[index]
        if price == ans:
            free_apps.append(app)
    return free_apps

android_free = isolate_free_app(android_english, 7, '0')
ios_free = isolate_free_app(ios_english, 4, '0.0')

print("android free apps number:", len(android_free))
print("ios free apps number:", len(ios_free))
    

android free apps number: 8864
ios free apps number: 0


### Data Anaysis
As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

We'll need to build a frequency table for the `prime_genre` column of the App Store data set, and for the `Genres` and `Category` columns of the Google Play data set.

We'll build two functions we can use to analyze the frequency tables:

- One function to generate frequency tables that show percentages
- Another function we can use to display the percentages in a descending order

In [14]:
def freq_table(dataset, index):
    table = {}
    total = 0
    for app in dataset:
        total += 1
        value = app[index]
        
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages

android_genre = freq_table(android_free, 9)
android_category = freq_table(android_free, 1)
ios_prime_genre = freq_table(ios_free, 11)

In [15]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        
display_table(ios_free, 11)

We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).

In [16]:
display_table(android_free, 1) #Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:

In [17]:
display_table(android_free, 9) #Genre

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

In [18]:
table_display = []
for genre in ios_prime_genre:
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[11]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    
    key_val_tuple = (avg_n_ratings, genre)
    table_display.append(key_val_tuple)
    
table_sorted = sorted(table_display, reverse = True)
for entry in table_sorted:
    print(entry[1], ':', entry[0])


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [19]:
for app in ios_free:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [20]:
for app in ios_free:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

- Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

- Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

- Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

In the previous screen, we came up with an app profile recommendation for the App Store based on the number of user ratings. We have data about the number of installs for the Google Play market, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [21]:
for app in android_category:
    total = 0
    len_category = 0
    for app_free in android_free:
        category_app = app_free[1]
        if category_app == app:
            n_install = app_free[5]
            n_install = n_install.replace('+', '')
            n_install = n_install.replace(',', '')
            total += float(n_install)
            len_category += 1
    ave_n_install = total / len_category
    print(app, ave_n_install)

ART_AND_DESIGN 1986335.0877192982
AUTO_AND_VEHICLES 647317.8170731707
BEAUTY 513151.88679245283
BOOKS_AND_REFERENCE 8767811.894736841
BUSINESS 1712290.1474201474
COMICS 817657.2727272727
COMMUNICATION 38456119.167247385
DATING 854028.8303030303
EDUCATION 1833495.145631068
ENTERTAINMENT 11640705.88235294
EVENTS 253542.22222222222
FINANCE 1387692.475609756
FOOD_AND_DRINK 1924897.7363636363
HEALTH_AND_FITNESS 4188821.9853479853
HOUSE_AND_HOME 1331540.5616438356
LIBRARIES_AND_DEMO 638503.734939759
LIFESTYLE 1437816.2687861272
GAME 15588015.603248259
FAMILY 3695641.8198090694
MEDICAL 120550.61980830671
SOCIAL 23253652.127118643
SHOPPING 7036877.311557789
PHOTOGRAPHY 17840110.40229885
SPORTS 3638640.1428571427
TRAVEL_AND_LOCAL 13984077.710144928
TOOLS 10801391.298666667
PERSONALIZATION 5201482.6122448975
PRODUCTIVITY 16787331.344927534
PARENTING 542603.6206896552
WEATHER 5074486.197183099
VIDEO_PLAYERS 24727872.452830188
NEWS_AND_MAGAZINES 9549178.467741935
MAPS_AND_NAVIGATION 4056941.774193

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs: